### PR TITLE
Feature/16 support the first two endpoints

### PR DESCRIPTION
**Pull request by windoverwater**
Initial pass at getting the first three endpoints to function

### CODE REVIEW

File: Makefile
Comment (stratofax) [inline Makefile:1]: I don't know anything about Makefiles, so I have nothing of value to add here.

File: poetry.lock
Comment (stratofax) [inline poetry.lock:1]: Looks like you ran `poetry update` and moved all the versions up to the latest. Generally, I prefer to do this in a separate PR, unless there's a code-related reason to update a package.

File: requirements.txt
Comment (stratofax) [inline requirements.txt:1]: Curious about a few things:

- Do we need a `requirements.txt` file if we're using 'poetry' & `poetry.lock` to track requirements. The answer may be "yes" because of the requirements of your "super-repo" -- as noted, I'm curious.
- If we do need this file, did you generate it using poetry? I found and tested this command: `poetry export -f requirements.txt --without-hashes > requirements.txt` -- from [Python poetry how to generate a requirements.txt -  Medium](https://servidorandycode.medium.com/python-poetry-how-to-generate-a-requirements-txt-a4fdcee26fe3)

File: src/vtp/web/api/data/favicon.ico
Comment (stratofax) [inline src/vtp/web/api/data/favicon.ico:1]: We've got a favicon.ico now? Cool! I'm not sure where it will appear, though. I thought you needed to serve up an HTML file to see the favicon.

File: src/vtp/web/api/main.py [hunk 1]
```diff
@@ -22,19 +41,63 @@ async def root() -> dict:
...
 
 
+# Endpoint #2
 @app.get("/vote/{vote_store_id}")
 async def get_empty_ballot(vote_store_id: str) -> dict:
     """Return an empty ballot for a given Vote Store ID"""
+
+    if vote_store_id in vote_store_ids:
+        # Cet a (the) blank ballot from the backend
+        cbo = CastBallotOperation(
```
Comment (stratofax) [inline src/vtp/web/api/main.py:70]: As noted, all calls to the vtp.ops backend, through line 77, should move to a separate module. Then, this call can be packaged in a function (easiest) or a class.

File: src/vtp/web/api/main.py [hunk 2]
```diff
@@ -22,19 +41,63 @@ async def root() -> dict:
...
+        )
+        return {"blank-ballot": empty_ballot}
+    return {"error": "VoteStoreID not found"}
+
+
+# Endpoint #3
+@app.get("/vote/cast-ballot/{vote_store_id}")
+async def cast_ballot(vote_store_id: str, cast_ballot_json: dict = None) -> dict:
+    """
+    Uploads the ballot; the backend accepts it, merges it, and returns
```
Comment (stratofax) [inline src/vtp/web/api/main.py:86]: As noted before, move this backend logic to a separate module. The API really doesn't care what the backend does with the data, only whether it was successful or not.

If handing the completed ballot to the backend is successful, the VoteStoreID needs to be removed from the `vote_store_ids` list so the ID can't be reused to vote again.

File: src/vtp/web/api/main.py [hunk 3]
```diff
@@ -22,19 +41,63 @@ async def root() -> dict:
...
+            verbosity=VERBOSITY,
+        )
+        # ZZZ stub out cast ballot
+        ballot_check, voter_index = abo.run(
+            cast_ballot=DEFAULT_CAST_BALLOT,
+            cast_ballot_json=cast_ballot_json,
+            merge_contests=True,
+        )
+        return {"ballot_check": ballot_check, "voter_index": voter_index}
+    return {"error": "VoteStoreID not found"}
```
Comment (stratofax) [inline src/vtp/web/api/main.py:103]: Assuming you'll move most of the code to that external vtp ops module, again, I'd suggest combining the if statement with the return in a one-liner, as in the previous endpoint.

File: src/vtp/web/api/json_data_models.py
Comment (stratofax) [inline src/vtp/web/api/json_data_models.py:1]: I'm not going to review this yet, because as noted in my comments on `main.py` I'd like you to move all the vtp.ops dependencies to a separate module. Then, it would make sense to test this class with the vtp.ops code.

File: poetry.lock
Comment (windoverwater) [reply poetry.lock:1]: oops - sorry.  Will do it that way from now on.

**Review by stratofax — CHANGES_REQUESTED** [review]
Please refactor `main.py` to move all of the vtp.ops dependencies to a separate module, as described in the  comments.

File: tests/test_main.py
Comment (stratofax) [inline tests/test_main.py:1]: Thanks for adding the missing docstrings! Much appreciated.

File: src/vtp/web/api/data/favicon.ico
Comment (windoverwater) [reply src/vtp/web/api/data/favicon.ico:1]: Er, I just stole one from the osetinstitute.org web site ...

File: requirements.txt
Comment (windoverwater) [reply requirements.txt:1]: GitHub still needs a requirements.txt file for dependabot and maybe other python checking/action stuff (TBD).  And yes that command is in the Makefile - I added a Makefile boilerplate to run that and a few other things.

File: Makefile
Comment (windoverwater) [reply Makefile:1]: Understood - it is just boilerplate to avoid bash or something else

File: src/vtp/web/api/main.py [hunk 4]
```diff
@@ -22,19 +41,63 @@ async def root() -> dict:
...
+        empty_ballot = cbo.run(
+            blank_ballot=DEFAULT_BLANK_BALLOT,
+            return_bb=True,
+        )
+        return {"blank-ballot": empty_ballot}
+    return {"error": "VoteStoreID not found"}
+
+
+# Endpoint #3
+@app.get("/vote/cast-ballot/{vote_store_id}")
```
[the remark this answers is not in the document]
Comment (windoverwater) [reply src/vtp/web/api/main.py:83]: Yes.  I figured that out but did not push that change because I got stuck with pydandic and typing trying to get the POST to work prior to commit'ing and pushing.

File: src/vtp/web/api/main.py [hunk 5]
```diff
@@ -1,19 +1,38 @@
...
 
 app = FastAPI()
 
-# create a list to store VoteStoreIDs
+########
+# backend demo constants
+########
+# where the ElectionData repo is
+DEFAULT_ELECTION_DATA_DIR = (
+    "/opt/VoteTrackerPlus/demo.01/mock-clients/scanner.00/VTP-mock-election.US.13"
```
[the remark this answers is not in the document]
Comment (windoverwater) [reply src/vtp/web/api/main.py:18]: Yes, but I think we need to talk about this on Tuesday.  It is possible that the path can go away if we remove some flexibility from the configurability of the demo.  Regardless I was going to do that after the demo but if we need to pull it into scope, I can do that.

File: src/vtp/web/api/main.py
[the remark this answers is not in the document]
Comment (windoverwater) [reply src/vtp/web/api/main.py:1]: Got it.  I think.  Regardless, will attempt refactoring.

File: src/vtp/web/api/main.py (hunk 2 above)
Comment (windoverwater) [reply src/vtp/web/api/main.py:86]: But the client needs the same ID to verify the ballot_check and tally the contest.  ?

File: src/vtp/web/api/main.py (hunk 3 above)
Comment (windoverwater) [reply src/vtp/web/api/main.py:103]: Yes.  Question, when you say an external vtp ops module, do you mean a file in this repo or a file in the VoteTrackerPlus repo?

File: src/vtp/web/api/json_data_models.py
Comment (windoverwater) [reply src/vtp/web/api/json_data_models.py:1]: Ok.  But to me, to move it seems to imply that I understand stuff about pydandic that I do not.  And I would like to understand sooner then later a few basic questions, like:

- should this be using pydandic and typing?  If not, should it use something else?
- if we are using pydandic and typing, do Lists need the length to be specified?  Ditto for the length of a Dict?  It seems as if I can understand the answers or understand why the question is a bad question (!), I may learn something :-) 

It would be good to understand this first so that I understand the general landscape.  Like, is the point here to place the pydandic code in the VoteTrackerPlus repo or the web-api repo?

File: src/vtp/web/api/main.py
[the remark this answers is not in the document]
Comment (ion-oset) [reply src/vtp/web/api/main.py:1]: @stratofax, @windoverwater I'm not sure this refactoring is necessary. FastAPI routes/endpoints are designed to be tested independently of each other using the [`TestClient`](https://fastapi.tiangolo.com/tutorial/testing/). You can also [test internals such as the database](https://fastapi.tiangolo.com/advanced/testing-database/), using PyTest fixtures to do dependency injection (see e.g. [How to set up and tear down a database between tests in FastAPI?](https://stackoverflow.com/questions/67255653/how-to-set-up-and-tear-down-a-database-between-tests-in-fastapi). The only thing that's not testable using that framework is the backend `vtp.ops` module - but the tests for that should go into the VTP core / VoteTrackerPlus.

Put another way: the semantics of the internals should be handled by the core independently of whether VTP is being used as a web API or not. If, and only if, that's not possible and there are _application semantics_ (not data representation or database access) that change when using a web API, should another layer be needed in the web service.

If there *is* some reason to split out the bodies of the endpoint functions further, note that ops code in this repository should _not_ go in the package `vtp.ops`. It's going to collide with the package in the VTP core. It's OK to call it `vtp.web.api.ops` but a more conventional name for a module1 with API endpoints would be `routes` or `endpoints`.

File: src/vtp/web/api/main.py [hunk 6]
```diff
@@ -1,19 +1,38 @@
...
-import random
 
-from fastapi import FastAPI
-from importlib_resources import files
+# import json
+import os
 
-import vtp.web.api.data
+from fastapi import FastAPI
+from vtp.ops.accept_ballot_operation import AcceptBallotOperation
```
[the remark this answers is not in the document]
Comment (ion-oset) [reply src/vtp/web/api/main.py:7]: Also as noted, this is probably unnecessary.

File: src/vtp/web/api/main.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ion-oset) [reply src/vtp/web/api/main.py:18]: > Also, all of these constants should go to the yet-to-be-created "vtp access module" described above

Also 👍🏼 with isolating program wide settings, no matter where they are set, in a single configuration module or file.

File: src/vtp/web/api/main.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ion-oset) [reply src/vtp/web/api/main.py:83]: Further reference: [RFC 9110 - GET](https://www.rfc-editor.org/rfc/rfc9110.html#name-get)

> Although request message framing is independent of the method used, content received in a GET request has no generally defined semantics, cannot alter the meaning or target of the request, and might lead some implementations to reject the request and close the connection because of its potential as a request smuggling attack ([Section 11.2](https://www.rfc-editor.org/rfc/rfc9112#section-11.2) of [[HTTP/1.1](https://www.rfc-editor.org/rfc/rfc9110.html#HTTP11)]). A client SHOULD NOT generate content in a GET request unless it is made directly to an origin server that has previously indicated, in or out of band, that such a request has a purpose and will be adequately supported. An origin server SHOULD NOT rely on private agreements to receive content, since participants in HTTP communication are often unaware of intermediaries along the request chain.

File: src/vtp/web/api/main.py [hunk 7]
```diff
@@ -22,19 +41,63 @@ async def root() -> dict:
...
+        cbo = CastBallotOperation(
+            guid=vote_store_id,
+            verbosity=VERBOSITY,
+        )
+        empty_ballot = cbo.run(
+            blank_ballot=DEFAULT_BLANK_BALLOT,
+            return_bb=True,
+        )
+        return {"blank-ballot": empty_ballot}
+    return {"error": "VoteStoreID not found"}
```
[the remark this answers is not in the document]
Comment (ion-oset) [reply src/vtp/web/api/main.py:79]: That's a general question worth answering: do we favor starting API endpoints with the success case since it's likely to be common, or get all error checking out of the way first? Both are reasonable, but it will pay off for reviewing and debugging if we are consistent.

File: src/vtp/web/api/main.py (hunk 2 above)
Comment (ion-oset) [reply src/vtp/web/api/main.py:86]: Then have a state attached to a given ID indicating whether it's been voted, verified, or tallied. The internals of that can be whatever makes sense.

**Review by ion-oset — COMMENTED** [review]
Initial review. Some comments on typos and a missing `make` target.

Overall looks good. However see my comment about the refactoring, which I don't think is necessary because of the support FastAPI  has for testing.

File: requirements.txt
Comment (stratofax) [reply requirements.txt:1]: This all sounds good! Thanks for the 411.

File: src/vtp/web/api/data/favicon.ico
Comment (stratofax) [reply src/vtp/web/api/data/favicon.ico:1]: Does it actually appear anywhere?

File: src/vtp/web/api/main.py
[the remark this answers is not in the document]
Comment (stratofax) [reply src/vtp/web/api/main.py:1]: I want to clarify that by "another module" I mean specifically "another Python file in this repo."

File: src/vtp/web/api/main.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (stratofax) [reply src/vtp/web/api/main.py:7]: Assuming we're all talking about the same thing (another .py file in this same repo), I like the idea of keeping the actual API endpoints file as simple as possible, meaning, with the minimum number of dependencies. The specific goal is to structure the code so we don't care whether we get the data from the vtp.ops backend, or a json file somewhere in the file system. It makes testing so much simpler.

File: src/vtp/web/api/main.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (stratofax) [reply src/vtp/web/api/main.py:79]: My philosopy is this: if we can avoid executing code, we should do it. If we can bounce out of a function early because of an error state, let's bounce! Thus: error check first.

File: src/vtp/web/api/main.py (hunk 2 above)
Comment (stratofax) [reply src/vtp/web/api/main.py:86]: @windoverwater We could continue to store the ID on the client for as long as needed. But, I don't think the client should be able to request **another** ballot if they've already cast one with a specific ID. I think @ion-oset is proposing a more sophisticated data structure than a simple list to prevent double-voting, but I'm not sure we need to store the ID on the server after the vote is cast and recorded. Let's figure this out soon, though.

File: src/vtp/web/api/main.py (hunk 3 above)
Comment (stratofax) [reply src/vtp/web/api/main.py:103]: As noted, a file in this repo.

File: src/vtp/web/api/json_data_models.py
Comment (stratofax) [reply src/vtp/web/api/json_data_models.py:1]: This suggestion makes sense to me.

File: src/vtp/web/api/main.py
[the remark this answers is not in the document]
Comment (stratofax) [reply src/vtp/web/api/main.py:1]: Also, I envisioned the following scenario: if the data access routines, like getting a new empty ballot, or recieving a cast ballot, are contained in a separate .py file in this repo, we could:

- Write tests for the data access routines that are independent of the api endpoints
- Have one version of this module on one branch that fetches the data from a local JSON file, and another that fetches the data from the vtp backend on another branch. This means any potential problems with the vtp backend code (not saying there will be any problems, of course, but you never know) won't break the API endpoints because we can stub in local JSON data on another branch. 

That's really my goal behind the refactoring: separate access to the vtp.ops backend from the API endpoints so they can be tested separately, then integrated when we know both are working the way we want.

[the remark this answers is not in the document]
Comment (ion-oset) [reply src/vtp/web/api/main.py:1]: TL;DR: Use PyTest Mocks. Don't add another layer to the source package _just_ for testing.

I'm sure you know that I am in favor of more abstraction rather than less for modularity's sake, and that includes for testing. But I don't think we need (or should) add that extra layer in the API source. We should do that only in the tests.

> Write tests for the data access routines that are independent of the api endpoints

Yes. But as those data access routines are in the [`vtp.ops`](https://github.com/TrustTheVote-Project/VoteTrackerPlus/tree/main/src/vtp/ops) module those tests belong in  [VoteTrackerPlus](https://github.com/TrustTheVote-Project/VoteTrackerPlus/) not in this package.

> Have one version of this module on one branch that fetches the data from a local JSON file, and another that fetches the data from the vtp backend on another branch. This means any potential problems with the vtp backend code (not saying there will be any problems, of course, but you never know) won't break the API endpoints because we can stub in local JSON data on another branch.

Agreed. But that kind of testing around a core dependency is precisely what mocks are for. Use [PyTest mock](https://pytest-mock.readthedocs.io/en/latest/usage.html) to mock out the `vtp.ops` access to return whatever we need for the tests, and you can test against JSON data, or any other data source without writing another layer in the package source.

[the remark this answers is not in the document]
Comment (ion-oset) [reply src/vtp/web/api/main.py:1]: > I want to clarify that by "another module" I mean specifically "another Python file in this repo."

Yes, understood. That's what "module" means in Python so I assumed that's what you meant.

File: src/vtp/web/api/main.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ion-oset) [reply src/vtp/web/api/main.py:7]: > It makes testing so much simpler.

It does. But you can get exactly the same effect with an even simpler implementation with test mocks.

File: src/vtp/web/api/main.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ion-oset) [reply src/vtp/web/api/main.py:79]: I'm inclined to agree, and am 👍🏼 on that. I want to clarify something I failed to say explicitly though, which is the use case where the reverse makes sense: when the the vast majority of calls will succeed, and the error checking is complex enough to cause a performance hit.

I don't think this is likely to happen here and definitely shouldn't happen for the demo.

File: src/vtp/web/api/main.py (hunk 2 above)
Comment (windoverwater) [reply src/vtp/web/api/main.py:86]: Note - I don't have me head around the bigger issues here yet, but regarding this one snippet, for the client to verify the ballot check (to provide E2EV) the backend needs to run git commands and those commands should only be run in the specific ID (guid git workspace) that was created for that specific voter.  This is due to the way we designed the 5 endpoints (which I think of as a good thing).  Without redesigning the current endpoints, which I would prefer not to do as that may cause new backend code to be written - which is more refactoring - endpoints 4 and 5 need to specify an ID (a guid) in which to run the git commands.  Hope this helps.

And yes, I lean towards a simple design where we just do not delete the guid.  It would be good to allow the voter to verify their ballot check many times as well as run the tally many times.  I don't understand where the ID is stored on the client (as in, is it stored in 'a browser tab' or is it stored 'in a browser'), but I am ok with the voter being able to vote multiple times or not being able to vote multiple times FWIIW.

File: src/vtp/web/api/data/favicon.ico
Comment (windoverwater) [reply src/vtp/web/api/data/favicon.ico:1]: I have yet to see it anywhere :-/

File: src/vtp/web/api/main.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (windoverwater) [reply src/vtp/web/api/main.py:7]: @ion-oset - do have the time to stub out some type of example?  Particularly if you have the time prior to when Neil and I meet in about 14 hours from now.  Understand that that is unlikely.

File: src/vtp/web/api/main.py [hunk 8]
```diff
@@ -1,19 +1,38 @@
...
+########
+# backend demo constants
+########
+# where the ElectionData repo is
+DEFAULT_ELECTION_DATA_DIR = (
+    "/opt/VoteTrackerPlus/demo.01/mock-clients/scanner.00/VTP-mock-election.US.13"
+)
+# pylint: disable=line-too-long
+DEFAULT_BLANK_BALLOT = "GGOs/states/Massachusetts/GGOs/towns/Concord/blank-ballots/json/000,001,002,003,ballot.json"
+# for testing only
```
[the remark this answers is not in the document]
Comment (windoverwater) [reply src/vtp/web/api/main.py:22]: I think it will go away first as the backend probably wants to deal with it.  The reason it is here is that the client needs the backend to invoke some git commands before there is a client allocated workspace in which to run git commands.  This can go away in several ways
